src/ir/module-utils.h

```diff
@@ -497,6 +497,34 @@ inline void collectHeapTypes(Module& wasm,
         counts.note(curr->type);
       } else if (curr->is<RttCanon>() || curr->is<RttSub>()) {
         counts.note(curr->type.getRtt().heapType);
+      } else if (auto* make = curr->dynCast<StructNew>()) {
+        // Some operations emit a HeapType in the binary format, if they are
+        // static and not dynamic (if dynamic, the RTT provides the heap type).
+        if (!make->rtt && make->type != Type::unreachable) {
+          counts.note(make->type.getHeapType());
+        }
+      } else if (auto* make = curr->dynCast<ArrayNew>()) {
+        if (!make->rtt && make->type != Type::unreachable) {
+          counts.note(make->type.getHeapType());
+        }
+      } else if (auto* make = curr->dynCast<ArrayInit>()) {
+        if (!make->rtt && make->type != Type::unreachable) {
+          counts.note(make->type.getHeapType());
+        }
+      } else if (auto* cast = curr->dynCast<RefCast>()) {
+        if (!cast->rtt && cast->type != Type::unreachable) {
+          counts.note(cast->getIntendedType());
+        }
+      } else if (auto* cast = curr->dynCast<RefTest>()) {
+        if (!cast->rtt && cast->type != Type::unreachable) {
+          counts.note(cast->getIntendedType());
+        }
+      } else if (auto* cast = curr->dynCast<BrOn>()) {
+        if (cast->op == BrOnCast || cast->op == BrOnCastFail) {
+          if (!cast->rtt && cast->type != Type::unreachable) {
+            counts.note(cast->getIntendedType());
+          }
+        }
```

Comment:
Good idea, I'll add small modules that verify we write out heap types of the new instructions here.

Otherwise, the new instructions should be fully tested in this PR already, so I'm not sure what you meant by the "as well" part?

Reply:
I meant testing just this particular functionality for both old and new instructions.

Reply:
Ok, then that should now be done.

(Note that I only added the cases where this can be tested, like a `struct.new`, as a `br*` will target a block/loop, where the heap type will also appear, so it's not possible to test for a crash there.)